driver.go

```
@@ -131,6 +131,133 @@ func (s *Session) processTriple(tx neo4j.Transaction, edge graph.Edge, nodeIds m
+		StartId: ob1.StartId,
+		EndId:   ob1.EndId,
+		Type:    ob1.Type,
+		Props:   ob1.Props,
+	}
+	return ob2
+}
+
+// there may be more than one record returned
+func findNeighbors(tx neo4j.Transaction, ids []int64) ([]neo4jNode, []neo4jNode, []neo4jEdge, error) {
+	sources := make([]neo4jNode, 0)
+	targets := make([]neo4jNode, 0)
+	edges := make([]neo4jEdge, 0)
+	// sl, _ := json.Marshal(ids)
+	// query := fmt.Sprintf("MATCH (n)-[e]->(m) where id(n) in %v return n,m,e", string(sl))
+	idrec, err := tx.Run("MATCH (n)-[e]->(m) where id(n) in $id return n,m,e", map[string]interface{}{"$id": ids})
+	if err != nil {
+		return sources, targets, edges, err
+	}
+	for idrec.Next() {
+		record := idrec.Record()
+		source := record.Values[0].(neo4j.Node)
+		sources = append(sources, newNode(source))
+		targets = append(targets, newNode(record.Values[1].(neo4j.Node)))
+		edges = append(edges, newEdge(record.Values[2].(neo4j.Relationship)))
+	}
+	return sources, targets, edges, nil
+}
+
+const (
+	shortEntitySchemaTerm = "ls:entitySchema"
+)
+
+func MakeProperties(input map[string]interface{}) map[string]*ls.PropertyValue {
+	ret := make(map[string]*ls.PropertyValue)
+	for k, v := range input {
+		switch v.(type) {
+		case string:
+			ret[k] = ls.StringPropertyValue(v.(string))
+		case []interface{}:
+			ret[k] = ls.StringSlicePropertyValue(v.([]string))
+		}
+	}
+	return ret
+}
+
+func loadEntityNodes(tx neo4j.Transaction, rootIds []int64, config Config, loadNeighbors func(neo4j.Transaction, []int64) ([]neo4jNode, []neo4jNode, []neo4jEdge, error)) (graph.Graph, error) {
+	grph := ls.NewDocumentGraph()
+	if len(rootIds) == 0 {
+		return grph, nil
+	}
+	// neo4j IDs
+	visitedNode := make(map[int64]graph.Node)
+	queue := make([]int64, 0, len(rootIds))
+	for _, id := range rootIds {
+		queue = append(queue, id)
+	}
+	for len(queue) > 0 {
+		srcNodes, adjNodes, adjRelationships, err := loadNeighbors(tx, queue)
+		if err != nil || len(srcNodes) == 0 {
+			return grph, nil
```

Don't check srcNodes len. you are ignoring the err.